libraries/chain/controller.cpp

```diff
@@ -2796,19 +2796,16 @@ struct controller_impl {
 
    // thread safe, expected to be called from thread other than the main thread
    block_state_legacy_ptr create_block_state_i( const block_id_type& id, const signed_block_ptr& b, const block_header_state_legacy& prev ) {
-      bool hs_active = false;
-      if (!b->header_extensions.empty()) {
-         std::optional<block_header_extension> instant_finality_ext = b->extract_header_extension(instant_finality_extension::extension_id());
-#warning change to use instant_finality_ext https://github.com/AntelopeIO/leap/issues/1508
-         if (instant_finality_ext) {
-            const auto& ext = std::get<instant_finality_extension>(*instant_finality_ext);
-            hs_active = !!ext.new_proposer_policy;
-         }
-      }
-
+      uint32_t hs_lib = hs_irreversible_block_num.load();
+      const bool hs_active = hs_lib > 0;
       auto trx_mroot = calculate_trx_merkle( b->transactions, hs_active );
-      EOS_ASSERT( b->transaction_mroot == trx_mroot, block_validate_exception,
-                  "invalid block transaction merkle root ${b} != ${c}", ("b", b->transaction_mroot)("c", trx_mroot) );
+      if( b->transaction_mroot != trx_mroot ) {
+         // Call of create_block_state_i can happen right before hs_irreversible_block_num
+         // is set. Fall back to verify in the other way.
+         trx_mroot = calculate_trx_merkle( b->transactions, !hs_active );
+         EOS_ASSERT( b->transaction_mroot == trx_mroot, block_validate_exception,
+                     "invalid block transaction merkle root ${b} != ${c}", ("b", b->transaction_mroot)("c", trx_mroot) );
+      }
```

Comment:
This opens up an attack vector as you can create a block with invalid mroot that will pass this validation. I think we can go with your first implementation. There is a small race condition at time of activation, but even if that is tripped and the block is considered invalid, the node will attempt to sync the block in the future and succeed; that is what I meant by re-try.

Reply:
Thanks for your insight!